src/jemalloc.c

```diff
@@ -2211,6 +2204,20 @@ malloc_init_hard(void) {
 	    || background_thread_boot1(tsd_tsdn(tsd), b0get())) {
 		UNLOCK_RETURN(tsd_tsdn(tsd), true, true)
 	}
+	if (opt_hpa && background_thread_enabled()) {
+		/*
+		 * Initialization of a0->pa_shard.hpa_shard happen before
+		 * background_thread_boot1 was called, so value of
+		 * background_thread_enabled_state was false at the time of
+		 * hpa_shard->opts initialization and hence
+		 * hpa_shard->opts.deferral_allowed was set to false as well.
+		 * We need to fix it here for arena 0 to have correct options
+		 * set.
+		 */
+		arena_t *a0 = arena_get(tsd_tsdn(tsd), 0, false);
+		pa_shard_set_deferral_allowed(tsd_tsdn(tsd), &a0->pa_shard,
```

Comment:
Let's move the entire a0 HPA enablement here, i.e. pa_shard_enable_hpa. Having a0's HPA partially initialized w/ a temporary deferral_allowed feels risky, e.g. if the HPA shard internal initialization makes any decision based on the deferral_allowed state. It's cleaner to make it either not enabled, or fully enabled.